maro/simulator/scenarios/finance/sub_engines/stock/stock_business_engine.py

```diff
@@ -0,0 +1,270 @@
+import os
+
+from maro.event_buffer import EventBuffer
+from maro.simulator.scenarios.finance.abs_sub_business_engine import \
+    AbsSubBusinessEngine
+from maro.simulator.scenarios.finance.common import (
+    Action, DecisionEvent,
+    FinanceType, TradeResult, OrderMode, ActionType, ActionState
+)
+
+from maro.data_lib import BinaryReader
+from .frame_builder import build_frame
+
+from maro.simulator.scenarios.finance.account import SubAccount
+
+from .stock import Stock
+from .stock_trader import StockTrader
+
+
+class StockBusinessEngine(AbsSubBusinessEngine):
+    def __init__(
+            self, beginning_timestamp: int, start_tick: int, max_tick: int,
+            frame_resolution: int, config: dict, event_buffer: EventBuffer
+    ):
+        super().__init__(beginning_timestamp, start_tick, max_tick, frame_resolution, config, event_buffer)
+
+        self._stock_codes: list = None
+        self._stocks_dict: dict = None
+        self._stock_list: list = None
+        self._readers: dict = None
+        self._order_mode = OrderMode.market_order
+        self._trader = None
+        self._item_picker = None
+        self._account: SubAccount = None
+
+        self._action_scope_max = self._config["action_scope"]["max"]
+
+        self._init_reader()
+        self._init_trader(self._config)
+
+        self._pending_orders = []
+        self._canceled_orders = []
+
+        self._init_frame()
+        self._build_stocks()
+
+    @property
+    def finance_type(self):
+        return FinanceType.stock
+
+    @property
+    def account(self):
+        return self._account
+
+    @property
+    def frame(self):
+        return self._frame
+
+    @property
+    def snapshot_list(self):
+        return self._snapshots
+
+    @property
+    def name_mapping(self):
+        return {stock.index: code for code, stock in self._stocks_dict.items()}
+
+    def step(self, tick: int):
+        valid_stocks = []
+
+        for code, _ in self._readers.items():
+            for raw_stock in self._item_picker[code].items(tick):
+                if raw_stock is not None:
+                    # update frame by code
+                    stock: Stock = self._stocks_dict[code]
+                    stock.fill(raw_stock)
+                    if stock.is_valid:
+                        valid_stocks.append(code)
+
+        # append cancel_order event
+        decision_event = DecisionEvent(
+            tick, FinanceType.stock, -2, self.name, self._action_scope, action_type=ActionType.cancel_order
+        )
+        evt = self._event_buffer.gen_cascade_event(tick, DecisionEvent, decision_event)
+        self._event_buffer.insert_event(evt)
+        # append account event
+        decision_event = DecisionEvent(
+            tick, FinanceType.stock, -1, self.name, self._action_scope, action_type=ActionType.transfer
```

Comment:
remove if transfer not available in this scenario. Please check all files for similar problem.

Reply:
done